src/kudu/consensus/consensus_queue.cc

```diff
@@ -760,15 +771,28 @@ Status PeerMessageQueue::RequestForPeer(const string& uuid,
       }
       msg_refs->swap(messages);
     } else {
+      // For a proxy request, we populate a dummy request with just a single
+      // 'op' which contains the first index that needs to be sent to this peer.
+      // The proxy host will read from the log to populate as many messages
+      // as needed. This avoids overloading the leader with unecessary disk
+      // reads. Note that the proxy host has to set the
+      // "preceding_id" correctly for the request (the one set by the leader
+      // here is not correct since the Ops are never read from the log)
       vector<ReplicateRefPtr> proxy_ops;
-      for (const ReplicateRefPtr& msg : messages) {
-        ReplicateRefPtr proxy_op = make_scoped_refptr_replicate(new ReplicateMsg);
-        *proxy_op->get()->mutable_id() = msg->get()->id();
-        proxy_op->get()->set_timestamp(msg->get()->timestamp());
-        proxy_op->get()->set_op_type(PROXY_OP);
-        request->mutable_ops()->AddAllocated(proxy_op->get());
-        proxy_ops.emplace_back(std::move(proxy_op));
-      }
+      ReplicateRefPtr proxy_op = make_scoped_refptr_replicate(new ReplicateMsg);
+
+      // A dummy OpId with the next index that has to be sent to the peer. The
+      // term is not relevent and set to '0'
+      OpId next_opid = MakeOpId(0, peer_copy.next_index);
```

Comment:
are there any truncation and deduplication checks that we do on intermediate node, or do we relegate that to the final destination?

Most of these checks go message by message and look for the valid matching messages.

Reply:
I chatted w/ Vinay about this offline. Ignoring term on the leader and having the secondary do it makes me nervous. I think the burden of proof of safety is fairly high to make this change and we should only do it if we are certain it's safe (maybe do a TLA+ proof) as well as believe it's a necessary optimization.